dcs/src/main/java/org/trafodion/dcs/master/listener/ListenerWorker.java

```diff
@@ -86,39 +86,43 @@ public void run() {
         DataEvent dataEvent;
     
         while(true) {
-            // Wait for data to become available
-            synchronized(queue) {
-                while(queue.isEmpty()) {
-                    try {
-                        queue.wait();
-                    } catch (InterruptedException e) {
+            try {
+                // Wait for data to become available
+                synchronized(queue) {
+                    while(queue.isEmpty()) {
+                        try {
+                            queue.wait();
+                        } catch (InterruptedException e) {
+                        }
                     }
+                    dataEvent = queue.remove(0);
                 }
-                dataEvent = queue.remove(0);
-            }
-            SelectionKey key = dataEvent.key;
-            SocketChannel client = (SocketChannel) key.channel();
-            Socket s = client.socket();
-            ClientData clientData = (ClientData) key.attachment();
-            ListenerService server = dataEvent.server;
-            dataEvent.key = null;
-            dataEvent.server = null;
-            
-            switch (clientData.hdr.getOperationId()){
-                case ListenerConstants.DCS_MASTER_GETSRVRAVAILABLE:
-                    clientData = requestGetObjectRef.processRequest(clientData, s);
-                    break;
-                case ListenerConstants.DCS_MASTER_CANCELQUERY:
-                    clientData = requestCancelQuery.processRequest(clientData, s);
-                    break;
-                default:
-                    clientData = requestUnknown.processRequest(clientData, s);
-                    break;
+                SelectionKey key = dataEvent.key;
+                SocketChannel client = (SocketChannel) key.channel();
+                Socket s = client.socket();
+                ClientData clientData = (ClientData) key.attachment();
+                ListenerService server = dataEvent.server;
+                dataEvent.key = null;
+                dataEvent.server = null;
+
+                switch (clientData.hdr.getOperationId()){
+                    case ListenerConstants.DCS_MASTER_GETSRVRAVAILABLE:
+                        clientData = requestGetObjectRef.processRequest(clientData, s);
+                        break;
+                    case ListenerConstants.DCS_MASTER_CANCELQUERY:
+                        clientData = requestCancelQuery.processRequest(clientData, s);
+                        break;
+                    default:
+                        clientData = requestUnknown.processRequest(clientData, s);
+                        break;
+                }
+                // Return to sender
+                int requestReply = clientData.requestReply;
+                key.attach(clientData);
+                server.send(new PendingRequest(key, requestReply));
+            } catch (Exception e){
+                LOG.error("Unexpected Exception", e);
```

Comment:
in this case if not catch Exception, there will throw the exception and lead to dcsmaster thread died, then the client side will never get connection, and once there catch Exception the cancelConnection will set to true, and then requestReply will be REQUST_CLOSE, and client side will get response for connect failure, this will not influence other client side